src/utils/jwt.ts

```diff
@@ -0,0 +1,275 @@
+/*!
+ * Copyright 2021 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as validator from './validator';
+import * as jwt from 'jsonwebtoken';
+import { HttpClient, HttpRequestConfig, HttpError } from '../utils/api-request';
+import { Agent } from 'http';
+
+export const ALGORITHM_RS256: jwt.Algorithm = 'RS256' as const;
+
+// `jsonwebtoken` converts errors from the `getKey` callback to its own `JsonWebTokenError` type
+// and prefixes the error message with the following. Use the prefix to identify errors thrown
+// from the key provider callback.
+// https://github.com/auth0/node-jsonwebtoken/blob/d71e383862fc735991fd2e759181480f066bf138/verify.js#L96
+const JWT_CALLBACK_ERROR_PREFIX = 'error in secret or public key callback: ';
+
+const NO_MATCHING_KID_ERROR_MESSAGE = 'no-matching-kid-error';
+
+export type Dictionary = { [key: string]: any }
+
+export type DecodedToken = {
+  header: Dictionary;
+  payload: Dictionary;
+}
+
+export interface SignatureVerifier {
+  verify(token: string): Promise<void>;
+}
+
+interface KeyFetcher {
+  fetchPublicKeys(): Promise<{ [key: string]: string }>;
+}
+
+/**
+ * Class to fetch public keys from a client certificates URL.
```

Comment:
Unrelated to this PR, but do these comments now need `@internal` annotations?

Reply:
Most of the time it's not required. But as a best practice I think we will ask developers add this annotation to any APIs that are exported from their residing modules, but are not part of the public API. This makes sure the generated d.ts files remain small.